doc/neps/nep-0029-dtype-as-type.rst

```diff
@@ -0,0 +1,268 @@
+is a python class with the simple ``mro`` ``[np.dtype, object]``. Creating an
+instance of ``dtype`` *i.e.* ``a = np.dtype('int8')`` will result in a python
+object of type ``dtype``. The ``dtype`` obect instance has attributes, among
+them ``a.type``, which is a class object. Instantiating that class object
+``a.type(3)`` produces a numpy `scalar
+<http://www.numpy.org/devdocs/reference/arrays.scalars.html>`.
+
+This NEP proposes a different class heirarchy. Objects of ``np.dtype`` will
+become type objects with a heirarchical ``mro`` like scalars. They will support
+subclassing. A future NEP may propose that instantiating a dtype type object
+will produce a scalar refleting that dtype, but that is not a goal of this NEP.
+
+The changed dtype will:
+
+- facilitate extending dtypes, typically for things like categoricals, novel
+  representations like datetime or IP addresses, or adding attributes like
+  units.
+- Simplify the code around ``__repr__`` and method lookup.
+
+Overall Design
+--------------
+
+In pure python (without error checking)::
+
+    import numpy as np
+
+    class Dtype(type):
+
+        def __new__(cls, obj, *args, **kwargs):
+            if isinstance(obj, int):
+                return dtype_int_dict[obj]
+            elif isinstance(obj, type) and issubclass(obj, np.generic):
+                return dtype_scalar_dict[obj]
+            elif len(args) < 1:
+                # Dtype('int8') or Dtype('S10') or record descr
+                return create_new_descr(cls, obj, *args, **kwargs)
+            else:
+                return super().__new__(cls, obj, *args, **kwargs)
+
+        def __call__(self, *args, **kwargs):
+            return self.typeobj(*args, **kwargs)
+
+    class IntDtype(Dtype):
+        def __repr__(self):
+            if self is IntDescr:
+                return type.__repr__(self)
+            return 'dtype(%s%d)' %(self.kind, self.itemsize)
+
+    class GenericDescr(type, metaclass=Dtype):
+        pass
+
+    class IntDescr(GenericDescr, metaclass=IntDtype):
+        def format(value):
+            return '%d' % value
+
+    class UInt8Descr(IntDescr):
+        kind = 'uint'
+        itemsize = 8
+        typeobj = np.uint8
+        # sort, fill, cast, clip, ...
+        ArrFuncs = int8_arrayfuncs
+
+    dtype_int_dict = {1: UInt8Descr}
+    dtype_scalar_dict = {np.uint8: UInt8Descr}
+
+At NumPy startup, as we do today, we would generate the builtin set of
+descriptor classes, and fill in ``dtype_int_dict`` and ``dtype_scalar_type``
+so that the built-in descriptors would continue to be singletons. ``Void``,
+``Byte`` and ``Unicode`` descriptors would be constructed on demand, as is done
+today.
+
+All dtype instances would inherit from ``GenericDescr`` which inherits from
+``type``, making them instances of ``type``::
+
+    >>> a = np.dtype(np.int8)
+    >>> a.mro(a)
+    [dtype(uint8), <class 'dtype.IntDescr'>, <class 'dtype.GenericDescr'>, \
```

Comment:
Note that the `mro` does not actually show the metaclass, because the metaclass methods are never looked up on instances. Though `isinstance(a, Dtype) is True`

Reply:
huh. Metaclasses are confusing. What is the deeper meaning of `isinstance(a, Dtype)` if methods are not inherited?

Reply:
It is indeed a bit confusing, and I am not sure why it is, though in our case, I think it is actually rather good that any `instance(array.dtype, np.dtype)` will continue to be true...